BUG: Fix case where DICOM series order has negative spacing

blowekamp opened the pull request:
If the spacing between slices is negative, then direction cosine matrix was not correct with reguards to the sign.

<!-- The text within this markup is a comment, and is intended to provide
guidelines to open a Pull Request for the ITK repository. This text will not
be part of the Pull Request. -->


<!-- See the CONTRIBUTING (CONTRIBUTING.md) guide. Specifically:

Start ITK commit messages with a standard prefix (and a space):

 * BUG: fix for runtime crash or incorrect result
 * COMP: compiler error or warning fix
 * DOC: documentation change
 * ENH: new functionality
 * PERF: performance improvement
 * STYLE: no logic impact (indentation, comments)
 * WIP: Work In Progress not ready for merge

Provide a short, meaningful message that describes the change you made.

When the PR is based on a single commit, the commit message is usually left as
the PR message.

A reference to a related issue or pull request (https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)
in your repository. You can automatically
close a related issues using keywords (https://help.github.com/articles/closing-issues-using-keywords/)

@mentions (https://help.github.com/articles/basic-writing-and-formatting-syntax/#mentioning-people-and-teams)
of the person or team responsible for reviewing proposed changes. -->

## PR Checklist
- [ ] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [ ] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [ ] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [ ] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [ ] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKSphinxExamples) for all new major features (if any)

Refer to the [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) for
further development details if necessary.

<!-- **Thanks for contributing to ITK!** -->

blowekamp commented:
relates to SimpleITK/SimpleITK#2292

dzenanz reviewed (approved):
Looks good after a quick look.

blowekamp commented:
I update the test case to use the sample MRI series in ITK test data, and use the ReverseOrder flag. The test is compare the pixel values at the physical locations match.

dzenanz commented:
Maybe we could have both of those tests?

blowekamp commented:
> Maybe we could have both of those tests?

The DICOM tags were not correct in the prior example test case where the DICOM files were manually created. And oddly the X,Y spacing was being swapped for some reason likely related to the TAGs. Also this test case for the file has a non identify cosine matrix, which helps  verifies the correct column was multiplied and not a row by -1.

N-Dekker commented:
Thanks @blowekamp I'm just trying to understand the issue. It reminds me of an ITKElastix issue reported by @thewtex:
- https://github.com/InsightSoftwareConsortium/ITKElastix/issues/291

Which appeared related to `ImageFileReader::GenerateOutputInformation()`: 
https://github.com/InsightSoftwareConsortium/ITK/blob/b375cefdc7fcd761a5581359e35dfecdc5f9d615/Modules/IO/ImageBase/include/itkImageFileReader.hxx#L222-L235

blowekamp commented:
The original issue was reported in SimpleITK here:
https://github.com/SimpleITK/SimpleITK/issues/2292

The cause appeared to be that the z-slice direction was negative. This was more easily reproduce by enabling the "ReverseOrder" flag.

It may very well be likely related to this change too:
https://github.com/InsightSoftwareConsortium/ITK/blob/b375cefdc7fcd761a5581359e35dfecdc5f9d615/Modules/IO/ImageBase/include/itkImageFileReader.hxx#L222-L235

It's not clear to me if the original DICOM reported in SimpleITK had negative spacing. The "ReverseOrder" reproduction of the issue, has negative inter-slice distance.  There is certainly areas that need further investigation.

The principle used in the test, that for DICOM if the "reverse order" flag is set then the physical location of the voxel should be maintained, should be correct. But DICOM is complicated.

issakomi commented:
For the the change related to _SpacingBetweenSlices_ s.
https://github.com/InsightSoftwareConsortium/ITK/issues/4794#issuecomment-2266073027
https://github.com/InsightSoftwareConsortium/ITK/issues/4794#issuecomment-2266616251
This may lead to issues if SpacingBetweenSlices is negative, surprisingly it is not uncommon (invalid from DICOM point of view, except for ancient NM storage).

DICOM doesn't force any order of slices. For classic series it is easy to sort files (ascending along slice normal, if IPP/IOP are available). This is done [in itkGDCMSeriesFileNames.cxx](https://github.com/InsightSoftwareConsortium/ITK/blob/624f745601b44d99c0d97b50eb846e1054f6fbc1/Modules/IO/GDCM/src/itkGDCMSeriesFileNames.cxx#L141).

For enhanced multi-frame instances is it not really done in GDCM, but if the order is descending (very very often) GDCM returns negative spacing and the code in [tkImageFileReader.hxx](https://github.com/InsightSoftwareConsortium/ITK/blob/b375cefdc7fcd761a5581359e35dfecdc5f9d615/Modules/IO/ImageBase/include/itkImageFileReader.hxx#L222-L235) (posted above) is triggered and so far the image is correct, but has left-handed matrix.

P.S. I don't use this stuff in my app, so it is just FYI.

issakomi commented:
> For enhanced multi-frame instances is it not really done in GDCM, but if the order is descending (very very often) GDCM returns negative spacing and the code in [tkImageFileReader.hxx](https://github.com/InsightSoftwareConsortium/ITK/blob/b375cefdc7fcd761a5581359e35dfecdc5f9d615/Modules/IO/ImageBase/include/itkImageFileReader.hxx#L222-L235) (posted above) is triggered and so far the image is correct, but has left-handed matrix.
> 

@N-Dekker To demonstrate how GDCM returns negative spacing and the code in [itkImageFileReader.hxx](https://github.com/InsightSoftwareConsortium/ITK/blob/b375cefdc7fcd761a5581359e35dfecdc5f9d615/Modules/IO/ImageBase/include/itkImageFileReader.hxx#L222-L235) is triggered:

The test file can be e.g. [this one](https://data.kitware.com/api/v1/file/5c193fcd8d777f072bd9a4e0/download), used for regular testing.

Add e.g. `std::cout << sp[0] << ' ' << sp[1] << ' ' << sp[2] << std::endl;` before `return` in [gdcmImageHelper.cxx GetSpacingValueFromSequence](https://github.com/InsightSoftwareConsortium/ITK/blob/624f745601b44d99c0d97b50eb846e1054f6fbc1/Modules/ThirdParty/GDCM/src/gdcm/Source/MediaStorageAndFileFormat/gdcmImageHelper.cxx#L317)
`1 1 -1.33`

or it is possible to load the file, save  as MHA and look at the line `ITK_original_spacing = 1 1 -1.33` (it is important to use recent ITK, there was a [bug](https://github.com/InsightSoftwareConsortium/ITK/commit/087a1c4fb0abe5fae279bf0dc285d4fb962424c6) in ITK_original_spacing).

blowekamp commented:
This patch does not directly deal with the DICOM tags, and specific DICOM information. It only correct a sign change in the DirectionCosine matrix when detected to maintain consistent physical spacing when the order is reversed. The specifics of multi-frame DICOM are not handled here.

If there is a specific DICOM series that should be tested, we can manually do that to ensure consistency.

Otherwise, I think this patch should be good?

blowekamp commented on Modules/IO/ImageBase/include/itkImageSeriesReader.hxx:
```diff
@@ -189,7 +189,14 @@ ImageSeriesReader<TOutputImage>::GenerateOutputInformation()
+      {
+        for (unsigned int j = 0; j < TOutputImage::ImageDimension; ++j)
+        {
+          direction[j][this->m_NumberOfDimensionsInImage] *= std::signbit(dirN[j]) ? -1.0 : 1.0;
```
Please clarify what you mean, the operation here is "*=" so only a sign correction is done.

blowekamp replied:
The existing test using the DICOM series data in ITK had a non-trivial direction cosine matrix, and was preserved.

Also I added more testing based on the discussion in the SimpleITK issue and explicitly create a DICOM series with negative spacing.

blowekamp commented:
> Thanks @blowekamp I'm just trying to understand the issue. It reminds me of an ITKElastix issue reported by @thewtex:
> 
> * [GroupwiseRegistration notebook: ERROR: the direction cosines matrix of the fixed image is invalid! ITKElastix#291](https://github.com/InsightSoftwareConsortium/ITKElastix/issues/291)
> 
> Which appeared related to `ImageFileReader::GenerateOutputInformation()`:
> 
> https://github.com/InsightSoftwareConsortium/ITK/blob/b375cefdc7fcd761a5581359e35dfecdc5f9d615/Modules/IO/ImageBase/include/itkImageFileReader.hxx#L222-L235

I just tries this patch with this dataset referenced. I used the file names in the order presented by `GDCMSeriesFileNames`:

    -00-slice009.dcm
    -00-slice008.dcm
    -00-slice007.dcm
    -00-slice006.dcm
    -00-slice005.dcm
    -00-slice004.dcm
    -00-slice003.dcm
    -00-slice002.dcm
    -00-slice001.dcm
    -00-slice000.dcm

When read in this order with the patch the direction(2,2) == -1.0, same as reported in the issue. When ReverseOrder is enable then direction(2,2,) == 1.0 and the physical geometry is the same. For both cases the space is `[0.976562, 0.976562, 2]`. 

This behavior is consistent and follow the expectations of the patch.

thewtex reviewed (dismissed):
Please see the inline comments.

Please also add a test where the image is not orthogonal to the z-plane, i.e. the third direction component is neither [0,0,1] or [0,0,-1].

blowekamp committed:
BUG: Fix case where DICOM series order has negative spacing

If the spacing between slices is negative, then direction cosine
matrix was not correct with reguards to the sign.

This can be easily created with the "ReverseOrder" flag with the
series reader and DICOM data. With the correction, a DICOM series
mantains its correct physical location when this flag is enabled.

hjmjohnson committed:
ENH: Tests — Extensive regression suite for PR 5357 direction-cosine fix

hjmjohnson commented:
@thewtex — the new test commit (`fcecee9e`) addresses both of your top-level requests directly. Quick visible summary:

- **"Add a test where the input does not have any 0's and 1's in the direction matrix"** → `GDCMSeriesDirection.Oblique_NonAxisAligned_Reversal` writes synthesized DICOM with row/col orientations derived from a non-trivial composed rotation (no zero/one elements), confirms forward and reversed file orderings give identical physical-point lookups for every voxel.
- **"Check that the resulting direction matrix's third vector equals the cross product of the first two"** → `GDCMSeriesDirection.ThirdColumnEqualsCrossProduct_Oblique` computes `c0 × c1` and asserts it is parallel to `c2` (sign-agnostic — ITK accepts left-handed direction matrices, det = ±1).

<details>
<summary>Full coverage map (17 gtest cases) and regression-detector check</summary>

| Test | Source / contract |
|---|---|
| `SimpleITK2292_BottomToTop_OriginAndDirectionAgree` / `SimpleITK2292_TopToBottom_OriginAndDirectionAgree` | [SimpleITK#2292](https://github.com/SimpleITK/SimpleITK/issues/2292) reproducer |
| `PhysicalPointInvariance_AxialReversal` / `PhysicalPointInvariance_ReverseOrderFlag` | Voxel-level invariance under file-order reversal |
| `Sagittal_AxisAligned_Reversal` / `Coronal_AxisAligned_Reversal` | Non-axial patient frames |
| `Oblique_NonAxisAligned_Reversal` | thewtex request #1 (no 0/1 entries) |
| `ThirdColumnEqualsCrossProduct_Oblique` | thewtex request #2 (cross-product check) |
| `ForceOrthogonalDirectionOff_UsesDirectIPPDifference` | Pin legacy code path |
| `ForceOrthogonalOnOff_AgreeOnAxisAligned` | On/off symmetry |
| `TwoSlices_MinimumValidCase` / `ManySlices_FineSpacing` | Boundary cases |
| `RealDicomSeries_PhysicalPointInvariant` | ITK's bundled `Input/DicomSeries` data |
| `ITKElastix291_TimeSeriesShapedDirection` | [ITKElastix#291](https://github.com/InsightSoftwareConsortium/ITKElastix/issues/291) contract |
| `OriginEqualsFirstListedSliceIPP` | Patch must not move origin |
| `SpacingAlwaysPositive` | Invariant |
| `RandomizedStress_PhysicalPointInvariance` | 8 random oblique trials, deterministic seed |

Reverting just the patched block in `itkImageSeriesReader.hxx` while keeping the suite causes **12 of the 17 tests to fail** (locally verified), so the suite is a faithful regression detector — not just exercises that happen to pass against the patched code.

</details>

<details>
<summary>Note on handedness</summary>

The patch produces left-handed direction matrices (det = −1) for canonical axial top-to-bottom DICOM where the IPP step is in −Z. ITK accepts this and the tests verify orthonormality with `|det| = 1` rather than `det = +1`. Worth flagging in case anyone reviewing thinks det should always be +1 — it's not, and the unpatched code already produces det = −1 in many DICOM cases.

</details>

<!--
provenance: claude-code session 2026-04-29 (hjmjohnson)
key_facts:
  test_file: Modules/IO/GDCM/test/itkGDCMSeriesDirectionGTest.cxx
  test_count: 17
  failures_when_patch_reverted: 12
  branch_tip: fcecee9e4c6af3281a7808d7a168a52ef583e1bf
  rebased_on: upstream/main 4b5eb76665 (2026-04-29)
  test_commit: ENH: Tests — Extensive regression suite for PR 5357 direction-cosine fix
related_files:
  - Modules/IO/GDCM/test/itkGDCMSeriesDirectionGTest.cxx
  - Modules/IO/GDCM/test/CMakeLists.txt:452-456
  - Modules/IO/ImageBase/include/itkImageSeriesReader.hxx:193-214
post_merge_action: none — patch + tests stand together
-->

dzenanz reviewed (commented):
I have taken a reasonable look at the first commit. It seems fine. I have not taken a good look at the tests in the second commit. Matt should review too.

hjmjohnson commented:
@dzenanz Thanks
@thewtex If you approve, would you please also merge?

thewtex reviewed (approved):
@hjmjohnson thank you!
